pallets/runtime/src/asset.rs

```diff
@@ -2020,4 +2082,96 @@ impl<T: Trait> Module<T> {
+            None => T::Balance::encode(&(0.into())),
+        };
+        let balance_from = match from_did {
+            Some(did) => T::Balance::encode(&<BalanceOf<T>>::get((ticker, &did))),
+            None => T::Balance::encode(&(0.into())),
+        };
+        let encoded_to = Option::<IdentityId>::encode(&to_did);
+        let encoded_from = Option::<IdentityId>::encode(&from_did);
+        let encoded_value = T::Balance::encode(&value);
+        let total_supply = T::Balance::encode(&<Tokens<T>>::get(&ticker).total_supply);
+
+        // Creation of the encoded data for the verifyTransfer function of the extension
+        // i.e fn verify_transfer(
+        //        from: Option<IdentityId>,
+        //        to: Option<IdentityId>,
+        //        value: Balance,
+        //        balance_from: Balance,
+        //        balance_to: Balance,
+        //        total_supply: Balance
+        //    ) -> RestrictionResult { }
+
+        let encoded_data = [
+            &selector[..],
+            &encoded_from[..],
+            &encoded_to[..],
+            &encoded_value[..],
+            &balance_from[..],
+            &balance_to[..],
+            &total_supply[..],
+        ]
+        .concat();
+
+        // Calling extension to verify the compliance rule
+        // native currency value should be `0` as no funds need to transfer to the smart extension
+        // We are passing arbitrary high `gas_limit` value to make sure extension's function execute successfully
+        // TODO: Once gas estimate function will be introduced, arbitrary gas value will be replaced by the estimated gas
+        let is_allowed =
+            Self::call_extension(extension_caller, dest, 0.into(), 5000000, encoded_data);
+        if is_allowed.is_success() {
+            if let Ok(allowed) = RestrictionResult::decode(&mut &is_allowed.data[..]) {
+                return allowed;
+            } else {
+                return RestrictionResult::Invalid;
+            }
+        } else {
+            return RestrictionResult::Invalid;
+        }
+    }
+
+    /// A helper function that is used to call the smart extension function.
+    ///
+    /// # Arguments
+    /// * `from` - Caller of the extension
+    /// * `dest` - Address/AccountId of the smart extension whom get called
+    /// * `value` - Amount of native currency that need to transfer to the extension
+    /// * `gas_limit` - Maximum amount of gas passed to successfully execute the function
+    /// * `data` - Encoded data that contains function selector and function arguments values.
+    pub fn call_extension(
+        from: T::AccountId,
+        dest: T::AccountId,
+        value: T::Balance,
+        gas_limit: Gas,
+        data: Vec<u8>,
+    ) -> ExecReturnValue {
+        // TODO: Fix the value conversion into Currency
+        match <pallet_contracts::Module<T>>::bare_call(from, dest, 0.into(), gas_limit, data) {
```

Comment:
Simplified: 
```
<pallet_contracts::Module<T>>::bare_call(from, dest, 0.into(), gas_limit, data)
   .map_err( |err| ExecReturnValue { status: 1, data: err.reason... });
```

Reply:
It can't use `map_err` as I am not using a Result as return type. Even If I do It will ask for two return values one is output another is an error but in our case, I am only returning an output and converting an error in the output as well.  If you think any other way to simplify it then please let me know.